rskj-core/src/main/java/co/rsk/core/bc/MiningMainchainViewImpl.java

```diff
@@ -34,79 +34,77 @@
     private Blockchain blockchain;
 
     @GuardedBy("internalBlockStoreReadWriteLock")
-    private Map<Keccak256, Block> blocksByHash;
+    private Map<Keccak256, BlockHeader> blocksByHash;
 
     @GuardedBy("internalBlockStoreReadWriteLock")
-    private Map<Long, List<Block>> blocksByNumber;
+    private Map<Long, List<BlockHeader>> blocksByNumber;
 
     @GuardedBy("internalBlockStoreReadWriteLock")
-    private List<Block> mainchain;
+    private List<BlockHeader> mainchain;
 
     public MiningMainchainViewImpl(Blockchain blockchain,
                                    int height) {
         this.height = height;
         this.blockchain = blockchain;
         this.blocksByHash = new HashMap<>();
         this.blocksByNumber = new HashMap<>();
-        fillInternalsBlockStore(blockchain.getBestBlock());
+        fillInternalStore(blockchain.getBestBlock().getHeader());
     }
 
-    @Override
-    public void addBestBlock(Block bestBlock) {
+    public void addBest(BlockHeader bestHeader) {
         synchronized (internalBlockStoreReadWriteLock) {
-            blocksByHash.put(bestBlock.getHash(), bestBlock);
-            addToBlockByNumberMap(bestBlock);
+            addHeaderToMaps(bestHeader);
+
+            BlockHeader previousBest = mainchain.get(0);
+            if (previousBest.isParentOf(bestHeader)) {
+                mainchain.add(0, bestHeader);
+            } else {
+                // new block is not on the current chain, so we need to recalculate it to keep it consistent
+                fillInternalStore(bestHeader);
```

Comment:
So we basically avoid recalculating the blockchain every time a block arrives right? Beisdes that is it correct recalculate the blockchain if the top is not the parent?

Reply:
Is this the memory leak that Juliano was worried?

Reply:
What would be ideal, in terms of BIg O performance, is to search the existent blockchain until a common parent is found. This is an extension of the case that checks if you are adding a block on the tip of the chain. 
Be careful because depending on the kind of reorganization, it may be also needed to add blocks at the end. That allows to complete the number of blocks the view holds as invariant.

Reply:
See 4705e52 for the changes